mayastor/src/bdev/nexus/nexus_child.rs

```diff
@@ -252,65 +259,112 @@ impl NexusChild {
+            Some(handle) => handle,
+            None => {
+                return Err(ChildError::ChildInvalid {});
+            }
+        };
 
         let block_size = bdev.block_len();
-
-        let primary = u64::from(block_size);
-        let secondary = bdev.num_blocks() - 1;
-
         let mut buf = desc
             .dma_malloc(block_size as usize)
             .context(LabelAlloc {})?;
 
-        self.read_at(primary, &mut buf)
+        self.read_at(0, &mut buf).await.context(LabelRead {})?;
+        let mbr = match Pmbr::from_slice(&buf.as_slice()[440 .. 512]) {
+            Ok(record) => record,
+            Err(_) => {
+                warn!(
+                    "{}: {}: The protective MBR is invalid!",
+                    self.parent, self.name
+                );
+                return Err(ChildError::LabelInvalid {});
+            }
+        };
+
+        self.read_at(u64::from(block_size), &mut buf)
             .await
             .context(LabelRead {})?;
+        let primary = match GPTHeader::from_slice(buf.as_slice()) {
+            Ok(header) => header,
+            Err(_) => {
+                warn!(
+                    "{}: {}: The primary GPT header is invalid!",
+                    self.parent, self.name
+                );
+                return Err(ChildError::LabelInvalid {});
+            }
+        };
 
-        let mut label = GPTHeader::from_slice(buf.as_slice());
-        if label.is_err() {
-            warn!(
-                "{}: {}: The primary label is invalid!",
-                self.parent, self.name
-            );
-            self.read_at(secondary, &mut buf)
-                .await
-                .context(LabelRead {})?;
-            label = GPTHeader::from_slice(buf.as_slice());
+        if mbr.entries[0].num_sectors != 0xffff_ffff
+            && mbr.entries[0].num_sectors as u64 != primary.lba_alt
+        {
+            warn!("{}: {}: The protective MBR disk size does not match the GPT disk size!", self.parent, self.name);
+            return Err(ChildError::LabelInvalid {});
         }
 
-        let label = match label {
-            Ok(label) => label,
-            Err(_) => return Err(ChildError::LabelInvalid {}),
+        self.read_at((bdev.num_blocks() - 1) * u64::from(block_size), &mut buf)
+            .await
+            .context(LabelRead {})?;
+        let secondary = match GPTHeader::from_slice(buf.as_slice()) {
+            Ok(header) => header,
+            Err(_) => {
+                warn!(
+                    "{}: {}: The secondary GPT header is invalid!",
+                    self.parent, self.name
+                );
+                return Err(ChildError::LabelInvalid {});
+            }
         };
 
+        if primary.guid != secondary.guid {
+            warn!("{}: {}: The primary and secondary GPT headers are inconsistent: GUIDs differ!", self.parent, self.name);
+            return Err(ChildError::LabelInvalid {});
```

Comment:
How do other tools handle primary vs secondary in case of a mismatch? Eg do they just defer to the secondary as the current source of truth?

Reply:
To be fair, this is something that should not ever happen. The spec requires that the secondary GPT must be updated whenever the primary changes, so that they are always in sync. If they disagree then (assuming the primary is valid) the secondary should be updated accordingly.
However my line of thinking was that, as the nexus requires identical labels accross all child disks, and the existing code will update all labels on all disks anyway whenever it finds at least one that is different, we can mark this as a failure and then the nexus will update everything and all will be well.
Granted this is not as efficient as it could be, but I was hoping to address that as part of CAS-146.